File: language-testing/udp_echo_server/udp_echo_server.rb
#! /usr/bin/env ruby

require 'socket'

def start_echo_server port
  socky = UDPSocket.new
  socky.bind("", port)
  loop do
    data, client = socky.recvfrom 1024
    _, client_port, client_host, _ = client
    socky.send data, 0, client_host, client_port

    File.open "udp_echo_server.log", "a" do |f|
      f.puts data
    end
  end
end

def die error
  $stderr.puts "Usage: #{$0} port\n#{error}"
  exit 1
end

if __FILE__ == $0
  die "No port given!" unless ARGV.length > 0
  die "Error: Port contains non-digits!" unless ARGV[0] =~ /^\d+$/
  start_echo_server ARGV[0].to_i
end

File: language-testing/udp_echo_server/udp_echo_server.pl
#! /usr/bin/env perl

use strict;
use warnings;

use IO::Handle;
use IO::Socket;
$| = 1;

sub start_echo_server {
  my $sock = new IO::Socket::INET (
    LocalPort => $_[0],
    Proto => "udp",
    Reuse => 1
  ) or die "Failed to start socket: $!\n";

  for (;;) {
    my $buf;
    $sock->recv($buf, 1024);
    $sock->send($buf);

    # Log the connection
    my $hostname = $sock->peerhost();
    open my $log, ">>", "udp_echo_server.log" or die "open: $!\n";
    print $log $buf;
    close $log;
  }
}

if ($#ARGV == 0) {
  die "Error: Port contains non-digits!" unless $ARGV[0] =~ /^\d+$/;
  &start_echo_server($ARGV[0]);
} else {
  print "Usage: $0 port\n";
  exit 1;
}
